staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -123,20 +124,42 @@ func newETCD3ReadyCheck(c storagebackend.Config, stopCh <-chan struct{}) (func()
 	return newETCD3Check(c, timeout, stopCh)
 }
 
+// atomic error acts as a cache for atomically store an error
+// the error is only updated if the timestamp is more recent than
+// current stored error.
+type atomicLastError struct {
+	mu        sync.RWMutex
+	err       error
+	timestamp time.Time
+}
+
+func (a *atomicLastError) Store(err error, t time.Time) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if a.timestamp.IsZero() || a.timestamp.Before(t) {
+		a.err = err
+		a.timestamp = t
+	}
+}
+
+func (a *atomicLastError) Load() error {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	return a.err
+}
+
 func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan struct{}) (func() error, error) {
```

Comment:
Do we construct one of these for every resource? Or one per backend?

Reply:
hmmmm https://github.com/kubernetes/kubernetes/pull/111559 :/

Reply:
the stack is

```
newETCD3Check
-- newETCD3ReadyCheck
-- newETCD3HealthCheck
---- CreateHealthCheck
---- CreateReadyCheck
------- addEtcdHealthEndpoint
-----------ApplyWithStorageFactoryTo
```

if I test sending multiple requests to the endpoints, doing this 

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
index 55e2cc0b42c..908f8717957 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go
@@ -168,6 +168,7 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
                        clientErr = fmt.Errorf("server is shutting down")
                }
        }()
+       counter := 0
 
        return func() error {
                // Given that client is closed on shutdown we hold the lock for
@@ -180,6 +181,8 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
                if clientErr != nil {
                        return clientErr
                }
+               counter++
+               fmt.Println("DEBUG ETCD CHECK", counter)
                ctx, cancel := context.WithTimeout(context.Background(), timeout)
                defer cancel()
                // See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
diff --git a/test/integration/apiserver/apiserver_test.go b/test/integration/apiserver/apiserver_test.go
index c371ead248f..9215f002262 100644
--- a/test/integration/apiserver/apiserver_test.go
+++ b/test/integration/apiserver/apiserver_test.go
@@ -210,6 +210,29 @@ func Test4xxStatusCodeInvalidPatch(t *testing.T) {
        }
 }
 
+func TestEtcdHealthz(t *testing.T) {
+       server := kubeapiservertesting.StartTestServerOrDie(t, nil, nil, framework.SharedEtcd())
+       defer server.TearDownFn()
+
+       rt, err := restclient.TransportFor(server.ClientConfig)
+       if err != nil {
+               t.Fatal(err)
+       }
+
+       for i := 0; i < 100; i++ {
+               req, err := http.NewRequest("GET", server.ClientConfig.Host+"/healthz/etcd", nil)
+               if err != nil {
+                       t.Fatal(err)
+               }
+               resp, err := rt.RoundTrip(req)
+               if err != nil {
+                       t.Fatal(err)
+               }
+               resp.Body.Close()
+
+       }
+}
```

the output only shows one "counter"

```
DEBUG ETCD CHECK 62
DEBUG ETCD CHECK 63
DEBUG ETCD CHECK 64
DEBUG ETCD CHECK 65
DEBUG ETCD CHECK 66
DEBUG ETCD CHECK 67
DEBUG ETCD CHECK 68
DEBUG ETCD CHECK 69
DEBUG ETCD CHECK 70
DEBUG ETCD CHECK 71
DEBUG ETCD CHECK 72
DEBUG ETCD CHECK 73
```

it seems one , I don't think that this does per backend, that seems to be done inside the client. that has the endpoints of the etcd backends

Reply:
That seems pretty promising, an even more conclusive check would be counting how many times newETCD3Check is called, maybe?

Reply:
I'm not totally convinced the testing and real startup sequences are sufficiently similar about this though 😬

Reply:
I can run in a real clusters though, so we are confident

Reply:
verified

```
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go
index 4c8a409d659..ed10ca7b513 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory.go
@@ -41,6 +41,7 @@ func Create(c storagebackend.ConfigForResource, newFunc func() runtime.Object) (
 
 // CreateHealthCheck creates a healthcheck function based on given config.
 func CreateHealthCheck(c storagebackend.Config, stopCh <-chan struct{}) (func() error, error) {
+       fmt.Println("CreateHealthCheck")
        switch c.Type {
```


```
kubectl -n kube-system logs kube-apiserver-test-control-plane | grep CreateHealthCheck | wc
      1       1      18
```